hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/config/validation/internal/EmulatesMapValidator.java

```diff
@@ -0,0 +1,118 @@
+import com.hedera.node.app.spi.config.validation.EmulatesMap;
+import com.swirlds.common.config.reflection.ConfigReflectionUtils;
+import com.swirlds.common.config.reflection.ConfigReflectionUtils.AnnotatedProperty;
+import com.swirlds.config.api.Configuration;
+import com.swirlds.config.api.validation.ConfigValidator;
+import com.swirlds.config.api.validation.ConfigViolation;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A {@link ConfigValidator} that validates that any property annotated with {@link EmulatesMap}. See {@link
+ * EmulatesMap} for more details.
+ */
+public class EmulatesMapValidator implements ConfigValidator {
+
+    @Override
+    @NonNull
+    public Stream<ConfigViolation> validate(@NonNull final Configuration configuration) {
+        Objects.requireNonNull(configuration, "Configuration cannot be null");
+        return ConfigReflectionUtils.getAllMatchingPropertiesForConstraintAnnotation(EmulatesMap.class, configuration)
+                .stream()
+                .map(this::convert)
+                .filter(Optional::isPresent)
+                .map(Optional::get);
+    }
+
+    @NonNull
+    private Optional<ConfigViolation> convert(@NonNull final AnnotatedProperty<EmulatesMap, ?> property) {
+        Objects.requireNonNull(property, "property cannot be null");
+        final Class<?> type = property.propertyType();
+        if (!List.class.isAssignableFrom(type) && !Set.class.isAssignableFrom(type)) {
+            final ConfigViolation violation =
+                    create(property, "Property is annotated with @EmulatesMap but is not a Set or List");
+            return Optional.of(violation);
+        }
+        final Type genericType = property.component().getGenericType();
+        if (genericType instanceof ParameterizedType parameterizedType) {
+            final Type singleGenericTypeArgument =
+                    ConfigReflectionUtils.getSingleGenericTypeArgument(parameterizedType);
+            if (!KeyValuePair.class.equals(singleGenericTypeArgument)) {
+                final ConfigViolation violation = create(
+                        property, "Property is annotated with @EmulatesMap but is not a Collection of KeyValuePair");
+                return Optional.of(violation);
+            }
+        } else {
+            final ConfigViolation violation =
+                    create(property, "Property is annotated with @EmulatesMap but is not a Collection of KeyValuePair");
+            return Optional.of(violation);
+        }
+        final var typedValue = (AnnotatedProperty<EmulatesMap, Collection<KeyValuePair>>) property;
+
+        final int uniqueKeyCount = typedValue.propertyValue().stream()
```

Comment:
How about using `Stream.distinct()`?

Reply:
But I need it to map to a collection anyway. How is that better than using set at the end?

Reply:
Why do you need to map to a collection? All you are interested in is the number of distinct elements, right? Would this work: `typedValue.propertyValue().stream().distinct().count()`?